cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -0,0 +1,195 @@
+              }
+            }
+          });
+        } finally {
+          ClassLoaders.setContextClassLoader(oldClassLoader);
+        }
+      } else {
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            delegate.onReceived(request.toByteBuffer());
+          }
+        });
+      }
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void finished(HttpResponder responder) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onFinish(BodyConsumerAdapter.this.responder);
+        }
+      });
+      BodyConsumerAdapter.this.responder.execute();
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void handleError(final Throwable cause) {
+    // When this method is called from netty-http, the response has already been sent, hence uses a no-op
+    // DelayedHttpServiceResponder for the onError call.
+    onError(cause, new DelayedHttpServiceResponder(responder) {
+      @Override
+      protected void doSend(int status, ByteBuffer content, String contentType,
+                            @Nullable Multimap<String, String> headers, boolean copy) {
+        // no-op
+      }
+
+      @Override
+      public void setTransactionFailureResponse(Throwable t) {
+        // no-op
+      }
+
+      @Override
+      public void execute() {
+        // no-op
+      }
+    });
+  }
+
+  /**
+   * Calls the {@link HttpContentConsumer#onError(HttpServiceResponder, Throwable)} method from a transaction.
+   */
+  private void onError(Throwable cause, final DelayedHttpServiceResponder responder) {
+    final Throwable failureCause = (cause instanceof WrappedException) ? cause.getCause() : cause;
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onError(responder, failureCause);
+        }
+      });
+    } catch (Throwable t) {
+      responder.setTransactionFailureResponse(t);
+    } finally {
+      responder.execute();
+    }
+  }
+
+  private Transactional createTransactional(final TransactionContext txContext,
+                                            final DatasetContext datasetContext) {
+    return new Transactional() {
+      @Override
+      public void execute(final TxRunnable runnable) {
+        try {
+          txContext.start();
+          try {
+            ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(contextClassLoader);
+            try {
+              runnable.run(datasetContext);
+            } finally {
+              ClassLoaders.setContextClassLoader(oldClassLoader);
+            }
+            txContext.finish();
+          } catch (Throwable t) {
```

Comment:
This implementation is adapted from `DefaultTransactionExecutor.executeOnce` method (https://github.com/caskdata/tephra/blob/develop/tephra-core/src/main/java/co/cask/tephra/DefaultTransactionExecutor.java#L131), which was used in the generated Service handler code as well prior to this change.

I just found a long standing tech debt related to this (https://issues.cask.co/browse/CDAP-617). Need to revisit the approach again and unify all code involved usage of `TransactionContext` directly.

For the wrapping, it is simply because the `Transactional.execute` don't have throws Exception in the signature (see above) so that there is a exception to unwrap it back when calling `onError`. Due to the refactoring of usage of `TransactionContext` is needed, the wrapping might goes away.